// opensea-js/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SHARED_STOREFRONT_LAZY_MINT_ADAPTER_CROSS_CHAIN_ADDRESS = exports.SHARED_STOREFRONT_ADDRESSES = exports.ENGLISH_AUCTION_ZONE_TESTNETS = exports.ENGLISH_AUCTION_ZONE_MAINNETS = exports.DEFAULT_ZONE = exports.API_BASE_TESTNET = exports.API_BASE_MAINNET = exports.MAX_EXPIRATION_MONTHS = exports.INVERSE_BASIS_POINT = exports.FIXED_NUMBER_100 = void 0;
const ethers_1 = require("ethers");
exports.FIXED_NUMBER_100 = ethers_1.FixedNumber.fromValue(100);
exports.INVERSE_BASIS_POINT = 10000n; // 100 basis points per 1%
exports.MAX_EXPIRATION_MONTHS = 1;
exports.API_BASE_MAINNET = "https://api.opensea.io";
exports.API_BASE_TESTNET = "https://testnets-api.opensea.io";
exports.DEFAULT_ZONE = ethers_1.ZeroAddress;
exports.ENGLISH_AUCTION_ZONE_MAINNETS = "0x110b2b128a9ed1be5ef3232d8e4e41640df5c2cd";
exports.ENGLISH_AUCTION_ZONE_TESTNETS = "0x9B814233894Cd227f561B78Cc65891AA55C62Ad2";
const SHARED_STOREFRONT_ADDRESS_MAINNET = "0x495f947276749ce646f68ac8c248420045cb7b5e";
const SHARED_STOREFRONT_ADDRESS_POLYGON = "0x2953399124f0cbb46d2cbacd8a89cf0599974963";
const SHARED_STOREFRONT_ADDRESS_KLAYTN = "0x5bc519d852f7ca2c8cf2d095299d5bb2d13f02c9";
exports.SHARED_STOREFRONT_ADDRESSES = [
    SHARED_STOREFRONT_ADDRESS_MAINNET,
    SHARED_STOREFRONT_ADDRESS_POLYGON,
    SHARED_STOREFRONT_ADDRESS_KLAYTN,
].map((address) => address.toLowerCase());
exports.SHARED_STOREFRONT_LAZY_MINT_ADAPTER_CROSS_CHAIN_ADDRESS = "0xa604060890923ff400e8c6f5290461a83aedacec";
