// @opensea/seaport-js/src/typechain-types/factories/@rari-capital/solmate/src/tokens/ERC721.sol/ERC721TokenReceiver__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  ERC721TokenReceiver,
  ERC721TokenReceiverInterface,
} from "../../../../../../@rari-capital/solmate/src/tokens/ERC721.sol/ERC721TokenReceiver";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
      {
        internalType: "address",
        name: "",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "",
        type: "bytes",
      },
    ],
    name: "onERC721Received",
    outputs: [
      {
        internalType: "bytes4",
        name: "",
        type: "bytes4",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ERC721TokenReceiver__factory {
  static readonly abi = _abi;
  static createInterface(): ERC721TokenReceiverInterface {
    return new Interface(_abi) as ERC721TokenReceiverInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): ERC721TokenReceiver {
    return new Contract(
      address,
      _abi,
      runner
    ) as unknown as ERC721TokenReceiver;
  }
}
