// @opensea/seaport-js/src/typechain-types/factories/src/contracts/test/TestERC1271Wallet.sol/IERC20Approve__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  IERC20Approve,
  IERC20ApproveInterface,
} from "../../../../../src/contracts/test/TestERC1271Wallet.sol/IERC20Approve";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "spender",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "approve",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IERC20Approve__factory {
  static readonly abi = _abi;
  static createInterface(): IERC20ApproveInterface {
    return new Interface(_abi) as IERC20ApproveInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): IERC20Approve {
    return new Contract(address, _abi, runner) as unknown as IERC20Approve;
  }
}
