// @opensea/seaport-js/src/typechain-types/factories/seaport_v1_5/contracts/lib/AmountDeriver__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  Contract,
  ContractFactory,
  ContractTransactionResponse,
  Interface,
} from "ethers";
import type { Signer, ContractDeployTransaction, ContractRunner } from "ethers";
import type { NonPayableOverrides } from "../../../../common";
import type {
  AmountDeriver,
  AmountDeriverInterface,
} from "../../../../seaport_v1_5/contracts/lib/AmountDeriver";

const _abi = [
  {
    inputs: [],
    name: "InexactFraction",
    type: "error",
  },
] as const;

const _bytecode =
  "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea26469706673582212201a9539d12cd5e2754332de8377ececf8e8fdcf55d447ec0bc03bc6c8c8e7c34164736f6c63430008110033";

type AmountDeriverConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AmountDeriverConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AmountDeriver__factory extends ContractFactory {
  constructor(...args: AmountDeriverConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override getDeployTransaction(
    overrides?: NonPayableOverrides & { from?: string }
  ): Promise<ContractDeployTransaction> {
    return super.getDeployTransaction(overrides || {});
  }
  override deploy(overrides?: NonPayableOverrides & { from?: string }) {
    return super.deploy(overrides || {}) as Promise<
      AmountDeriver & {
        deploymentTransaction(): ContractTransactionResponse;
      }
    >;
  }
  override connect(runner: ContractRunner | null): AmountDeriver__factory {
    return super.connect(runner) as AmountDeriver__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AmountDeriverInterface {
    return new Interface(_abi) as AmountDeriverInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): AmountDeriver {
    return new Contract(address, _abi, runner) as unknown as AmountDeriver;
  }
}
