// @opensea/seaport-js/src/typechain-types/factories/seaport_v1_5/contracts/lib/FulfillmentApplier__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  Contract,
  ContractFactory,
  ContractTransactionResponse,
  Interface,
} from "ethers";
import type { Signer, ContractDeployTransaction, ContractRunner } from "ethers";
import type { NonPayableOverrides } from "../../../../common";
import type {
  FulfillmentApplier,
  FulfillmentApplierInterface,
} from "../../../../seaport_v1_5/contracts/lib/FulfillmentApplier";

const _abi = [
  {
    inputs: [],
    name: "InvalidFulfillmentComponentData",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "fulfillmentIndex",
        type: "uint256",
      },
    ],
    name: "MismatchedFulfillmentOfferAndConsiderationComponents",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "enum Side",
        name: "side",
        type: "uint8",
      },
    ],
    name: "MissingFulfillmentComponentOnAggregation",
    type: "error",
  },
  {
    inputs: [],
    name: "OfferAndConsiderationRequiredOnFulfillment",
    type: "error",
  },
] as const;

const _bytecode =
  "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea264697066735822122074eb1f669f077f5e658bd2de4642408d84513b26b143734c8fc4359911bf6af964736f6c63430008110033";

type FulfillmentApplierConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: FulfillmentApplierConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class FulfillmentApplier__factory extends ContractFactory {
  constructor(...args: FulfillmentApplierConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override getDeployTransaction(
    overrides?: NonPayableOverrides & { from?: string }
  ): Promise<ContractDeployTransaction> {
    return super.getDeployTransaction(overrides || {});
  }
  override deploy(overrides?: NonPayableOverrides & { from?: string }) {
    return super.deploy(overrides || {}) as Promise<
      FulfillmentApplier & {
        deploymentTransaction(): ContractTransactionResponse;
      }
    >;
  }
  override connect(runner: ContractRunner | null): FulfillmentApplier__factory {
    return super.connect(runner) as FulfillmentApplier__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): FulfillmentApplierInterface {
    return new Interface(_abi) as FulfillmentApplierInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): FulfillmentApplier {
    return new Contract(address, _abi, runner) as unknown as FulfillmentApplier;
  }
}
